### DISCUSSION TITLE
Can't hide/remove console even with arguments

### DISCUSSION THREAD

**Discussion by Iuryck** [pyinstaller/pyinstaller#8777] [Help] [answered]
I am running the latest Pyinstaller version to deploy a Kivy app. I am using all possible arguments to remove the console, even adding the args to the spec files, but it still opens the console. I am using Git Actions and a Yaml file to do this deployments so it could be that, running on `windows-latest`

Here is my spec:

```
import os
os.environ['KIVY_GL_BACKEND'] = 'angle_sdl2'

from kivy_deps import sdl2, glew
from kivymd import hooks_path as kivymd_hooks_path

# -*- mode: python ; coding: utf-8 -*-

import argparse

parser = argparse.ArgumentParser()
parser.add_argument('--noconsole', action='store_true')
parser.add_argument('--windowed', action='store_true')
options = parser.parse_args()

block_cipher = None


a = Analysis(
    ['main.py'],
    pathex=[],
    binaries=[],
    datas=[('./pages','./pages'),('./assets','./assets'),('./utils','./utils')],
    hiddenimports=[],
    hookspath=[kivymd_hooks_path],
    hooksconfig={},
    runtime_hooks=[],
    excludes=[],
    win_no_prefer_redirects=False,
    win_private_assemblies=False,
    cipher=block_cipher,
    noarchive=False,
)
pyz = PYZ(a.pure, a.zipped_data, cipher=block_cipher)



exe = EXE(
    pyz,
    a.scripts,
    [],
    exclude_binaries=True,
    name='main',
    debug=False,
    bootloader_ignore_signals=False,
    strip=False,
    upx=True,
    console=True,
    disable_windowed_traceback=False,
    argv_emulation=False,
    target_arch=None,
    codesign_identity=None,
    entitlements_file=None,
    icon=['assets\\icon.ico'],
)
coll = COLLECT(
    exe,
    Tree('dist\\main'),
    a.binaries,
    a.zipfiles,
    a.datas,
    *[Tree(p) for p in (sdl2.dep_bins + glew.dep_bins)],
    strip=False,
    upx=True,
    upx_exclude=[],
    name='main',
    console=False
)

```

Here is my Yaml:

```
name: Build Recette App Desktop

on:
  push:
    branches:
      - staging
  pull_request:
    branches:
      - staging

jobs:
  build:
    runs-on: windows-latest

    steps:
    - name: Checkout code
      uses: actions/checkout@v4

    - name: Set up Python
      uses: actions/setup-python@v5
      with:
        python-version: '3.11.9'

    - name: Create Virtual Env
      run: |
        python -m pip install --upgrade pip
        pip install virtualenv
        mkdir my_env
        cd my_env
        python -m venv . 
        ./Scripts/Activate.ps1
        Write-Host "VIRTUAL_ENV: $env:VIRTUAL_ENV"
        echo "VIRTUAL_ENV=$VIRTUAL_ENV" >> $GITHUB_ENV

        

    - name: Install Dependencies
      run: |
        my_env/Scripts/Activate.ps1
        pip install -r requirements.txt

    - name: Set Kivy GL Backend
      run: |
        my_env/Scripts/Activate.ps1
        echo "KIVY_GL_BACKEND=angle_sdl2" >> $env:GITHUB_ENV
      

    - name: Build with PyInstaller
      run: |
        my_env/Scripts/Activate.ps1
        pyinstaller --windowed --noconsole --onedir --icon assets/icon.ico  main.py
      

    - name: Config main.spec
      run: |
        echo "${{ vars.spec_file }}" >> main.spec
      shell: bash
    
    - name: Recompile with .spec
      run: |
        my_env/Scripts/Activate.ps1
        pyinstaller main.spec -y    


    - name: Recette zip artifact
      uses: actions/upload-artifact@v4
      with:
        name: recette
        path: dist
        include-hidden-files: true
```

**Comment by bwoodsend** [comment]
Are you sure it's your application's console and not that of a subprocess? That catches people out quite often. e.g. https://github.com/pyinstaller/pyinstaller/issues/8748

**Reply by Iuryck** [reply]
Yeah I'm sure. It starts up showing all that Kivy log building up the app, it also shows some print statements I put in to measure building time. Different from the issue you mentioned, the console stays

**Comment by rokm** [comment, marked as answer]
>``` 
>exe = EXE(
>    ...
>    console=True,
>    ...
>)
>``` 

Surely setting `console=False` should be a good start?

Otherwise, I imagine that your `Build with PyInstaller` creates a noconsole build, and then you overwrite the spec (or rather, you seem to be appending to it?), and if your effective spec is what you have shown us, then `Recompile with .spec` overwrites the previous noconsole build with a console one.

**Reply by Iuryck** [reply]
Interesting, I did `console=False` inside `col = COLLECT` bc that was the one I found. I'll surely try this
Still shouldn't `--windowed --noconsole` be enough? All the docs say it is, and it was in previous versions

**Reply by rokm** [reply]
> Interesting, I did `console=False` inside `col = COLLECT` bc that was the one I found. I'll surely try this 

It needs to be in EXE; the one you put in COLLECT does nothing.

> Still shouldn't `--windowed --noconsole` be enough? All the docs say it is, and it was in previous versions

But you're not specifying `--windowed`/`--noconsole` in the second build, are you?

```
    - name: Recompile with .spec
      run: |
        my_env/Scripts/Activate.ps1
        pyinstaller main.spec -y 
```

So it uses whatever is in the .spec file.

But either way, the spec-generating flags (like `--windowed`/`--noconsole`) cannot be used when building with .spec file (because these settings are already defined in the .spec), so trying to use them with recent PyInstaller versions raises an error, while in older PyInstaller versions, they were quietly ignored.

**Reply by rokm** [reply]
And while we're at it, you should also remove the `Tree('dist\\main')` part from

```
coll = COLLECT(
    ...
    Tree('dist\\main'),
    ...
)
```

How did you even come up with this nonsense?

**Reply by Iuryck** [reply]
> 
> But you're not specifying `--windowed`/`--noconsole` in the second build, are you?
> 

Yeah... you can't specify that... it raises an error. Why are you even bringing it up?

Still I tried with 

```
parser = argparse.ArgumentParser()
parser.add_argument('--noconsole', action='store_true')
parser.add_argument('--windowed', action='store_true')
options = parser.parse_args()
```

**Reply by Iuryck** [reply]
> And while we're at it, you should also remove the `Tree('dist\\main')` part from
> 
> ```
> coll = COLLECT(
>     ...
>     Tree('dist\\main'),
>     ...
> )
> ```
> 
> How did you even come up with this nonsense?

That's because I'm building a DIR and not a single file. It collects all the needed files and puts it in the final DIR

**Reply by rokm** [reply]
> > But you're not specifying `--windowed`/`--noconsole` in the second build, are you?
> 
> Yeah... you can't specify that... it raises an error. Why are you even bringing it up?

Because the point is, --windowed/--noconsole works only when generating spec (e.g., generating it on-the-fly, like you do in your `Build with PyInstaller`).

> Still I tried with
> 
> ```
> parser = argparse.ArgumentParser()
> parser.add_argument('--noconsole', action='store_true')
> parser.add_argument('--windowed', action='store_true')
> options = parser.parse_args()
> ```

Yeah, that's another puzzling thing in your .spec - you create a command-line parser instance that accepts those options and supposedly parses the arguments, but you don't use resulting `options` anywhere in the spec file. But it's all moot anyway, because arguments passed to PyInstaller (which you are re-parsing here) cannot contain `--noconsole`/`--windowed` due to the above point.

**Reply by Iuryck** [reply]
> > Still I tried with
> > ```
> > parser = argparse.ArgumentParser()
> > parser.add_argument('--noconsole', action='store_true')
> > parser.add_argument('--windowed', action='store_true')
> > options = parser.parse_args()
> > ```
> 
> Yeah, that's another puzzling thing in your .spec - you create a command-line parser instance that accepts those options and supposedly parses the arguments, but you don't use resulting `options` anywhere in the spec file. But it's all moot anyway, because arguments passed to PyInstaller (which you are re-parsing here) cannot contain `--noconsole`/`--windowed` due to the above point.

Yeah I know it's weird. But it's in the docs so I tried it: https://pyinstaller.org/en/v6.2.0/spec-files.html#adding-parameters-to-spec-files

**Reply by rokm** [reply]
> That's because I'm building a DIR and not a single file. It collects all the needed files and puts it in the final DIR

This should never be required, unless you have some weird co-dependency going on between your first build step (`Build with PyInstaller`) and the second one (`Recompile with .spec`).

**Reply by rokm** [reply]
> > > Still I tried with
> > > ```
> > > parser = argparse.ArgumentParser()
> > > parser.add_argument('--noconsole', action='store_true')
> > > parser.add_argument('--windowed', action='store_true')
> > > options = parser.parse_args()
> > > ```
> > 
> > 
> > Yeah, that's another puzzling thing in your .spec - you create a command-line parser instance that accepts those options and supposedly parses the arguments, but you don't use resulting `options` anywhere in the spec file. But it's all moot anyway, because arguments passed to PyInstaller (which you are re-parsing here) cannot contain `--noconsole`/`--windowed` due to the above point.
> 
> Yeah I know it's weird. But it's in the docs so I tried it: https://pyinstaller.org/en/v6.2.0/spec-files.html#adding-parameters-to-spec-files

Ah, right, I forgot you can pass additional arguments after `--` and parse *those* instead. But this is not what you are doing - you do not use `options` in the spec, and you did not pass extra arguments to PyInstaller command after `--`).

**Reply by Iuryck** [reply]
> Ah, right, I forgot you can pass additional arguments after `--` and parse _those_ instead. But this is not what you are doing - you do not use `options` in the spec, and you did not pass extra arguments to PyInstaller command after `--`).

Yeah I knew something was wrong about it but Im in such a headache with the app itself I was just hoping to do some magic and just make the thing work.  

Anyway, the `console=False` inside EXE worked. I really really wished they just putt this stuff in all the quick starts and tutorials in the documentation..... Thanks
